Adds fast gradient clipping support for the Embedding layer.

pagarwl opened the pull request:
The algorithm used is described in the 'A Unified Fast Gradient Clipping Framework for DP-SGD' paper: https://proceedings.neurips.cc/paper_files/paper/2023/file/a45d344b28179c8da7646bc38ff50ad8-Paper-Conference.pdf. 
 
## Types of changes

<!--- What types of changes does your code introduce? Please mark an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

Previously, Ghost clipping was not supported in Opacus for embedding layer. With default DP-SGD implementation, the training OOMs out on large embedding layers over large physical batch size (useful for privacy). The regular DP-SGD needs O(Bnd), where B=physical batch size, n=vocab size, d=embedding dimension. To give an example on memory needed: we've seen embeddings with [vocab size=1000000, dim=5] (and higher) in real-world differential privacy applications. With a physical batch size of 16,000, memory needed: 16000 × 1000000 × 5 x 4 = 298.02 GB. 

With this change, we need significantly smaller memory: O(Br) where B is physical batch size, and r is number of unique  indices in the embedding sequence. We could successfully run DP-SGD over above example, using < 8GiB.

This is a good add to Opacus, enabling larger embedding layers over larger physical batch sizes to be trained with DP-SGD.

## How Has This Been Tested (if it applies)

Unit tests. Runs over large embedding layer training over realworld DP application.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

EnayatUllah commented:
Thank you for the PR! Overall, the logic look good to me -- I left some minor comments. Additionally, I suggest to merge the content of the file `embedding_norm_sample.py` to `embedding.py` itself.

EnayatUllah commented on opacus/tests/grad_sample_module_fast_gradient_clipping_test.py:
```diff
@@ -260,3 +276,151 @@ def test_gradient_calculation_fast_gradient_clipping(self, size, length, dim):
+        diff = flat_norms_normal - flat_norms_gc
+
+        logging.info(f"Diff = {diff}")
+        msg = "Fail: Gradients from vanilla DP-SGD and from fast gradient clipping are different"
```
msg should be "Fail: Gradient norms from vanilla DP-SGD and from fast gradient clipping are different"

pagarwl replied:
Done.

EnayatUllah commented on opacus/tests/grad_samples/embedding_norm_sample_test.py:
```diff
@@ -0,0 +1,218 @@
+        expected_norms = torch.tensor(
+            [0.0150, 0.0071, 0.0005, 0.0081, 0.0039], dtype=torch.float32
+        )
+        print("expected_norms: ", expected_norms)
```
Remove the print statement

pagarwl replied:
Done.

HuanyuZhang commented on opacus/grad_sample/embedding_norm_sample.py:
```diff
@@ -0,0 +1,150 @@
+    input_ids = input_ids.reshape(input_ids.shape[0], -1)
+
+    # Reshape grad_values preserving the embedding dimension as the last dimension
+    grad_values = grad_values.reshape(-1, grad_values.size(-1))
```
@pagarwl  just double checking, this implementation not only works for 2 dimensions ([batch_size, seq_length]), but also works for any general dimension, correct?

HuanyuZhang commented on opacus/tests/grad_samples/embedding_norm_sample_test.py:
```diff
@@ -0,0 +1,218 @@
+        # Manually set weights for the embedding layer for testing
+        embedding_layer.weight = nn.Parameter(
+            torch.tensor([[0.1], [0.2], [0.3]], dtype=torch.float32)
+        )
```
Nitpicking: although it will not be used in calculation, could we change the shape from [3,1] to [3,2] to match the embedding_dim=2? @pagarwl

pagarwl replied:
Done.

HuanyuZhang commented on opacus/tests/grad_samples/embedding_norm_sample_test.py:
```diff
@@ -0,0 +1,218 @@
+        )
+
+        # Example gradients per input id, with embedding_dim=2.
+        # Shape: [6, 1, 1, 2]
```
It seems the shape is in a different logic from the first example (``test_compute_embedding_norm_sample``), where input has shape [3,2] and ``grad_values`` has shape [6,1]. Do you mind clarifying?

HuanyuZhang commented on opacus/tests/grad_samples/embedding_norm_sample_test.py:
```diff
@@ -0,0 +1,218 @@
+        input_ids = torch.tensor([[1, 1], [2, 0], [2, 0]], dtype=torch.long)
+
+        # Example gradients with respect to the embedding output (backprops).
+        # Shape: [6, 1]
```
Nitpicking: we should have size [3,2,1] for backprops, correct? [6, 1] is for grad_values.

pagarwl replied:
yes, updated.

HuanyuZhang commented on opacus/tests/grad_samples/embedding_norm_sample_test.py:
```diff
@@ -0,0 +1,218 @@
+
+        # Example gradients per input id, with embedding_dim=2.
+        # Shape: [6, 1, 1, 2]
+        grad_values = torch.tensor(
```
Similar here, maybe we should distinguish between grad_values and backprops, since they have different shapes?

pagarwl replied:
Done.

HuanyuZhang commented on opacus/grad_sample/embedding_norm_sample.py:
```diff
@@ -0,0 +1,150 @@
+      activations:  [tensor([[1, 1],
+          [2, 0],
+          [2, 0]])]
+      backprops:  tensor([[0.2000],
```
Wrong shape for backprops, which should be [3,2,1]. [6,1] is for grad_values

pagarwl replied:
Done.

HuanyuZhang reviewed (commented):
Do you mind also fixing the lint error due to isort? https://github.com/pytorch/opacus/actions/runs/12342666465/job/34611577396

pagarwl commented:
> Do you mind also fixing the lint error due to isort? https://github.com/pytorch/opacus/actions/runs/12342666465/job/34611577396

Done.

pagarwl commented:
> Thank you for the PR! Overall, the implementation looks good to me -- I left some minor comments. Additionally, I suggest to merge the content of the file `embedding_norm_sample.py` to `embedding.py` itself.

Thanks for the review. I have incorporated all the changes. My colleague can pursue merging the contents of`embedding_norm_sample.py` to `embedding.py` in a follow-up PR. I've moved to another team in the last couple of months and I am pressed on time.

EnayatUllah commented:
Looks good to me. @pagarwl, can you fix the isort error? Run

```
isort -l 88 -o opacus --lines-after-imports 2 -m 3 --trailing-comma  .
```
from the repository root. Configuration for isort is located in .isort.cfg.
If using `isort` versions `<5.0.0` call
```bash
isort -l 88 -o opacus --lines-after-imports 2 -m 3 --trailing-comma --recursive
